mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java

```diff
@@ -0,0 +1,105 @@
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.elasticsearch;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.elasticsearch.action.ListenableActionFuture;
+import org.elasticsearch.action.bulk.BulkRequestBuilder;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.msgpack.core.Preconditions;
+
+public class DeleteByQueryPerformer {
+
+    public static final String DELETE_BY_QUERY_POOL_SIZE = "DeleteByQueryPoolSize";
+    public static final String DELETE_BY_QUERY_BATCH_SIZE = "DeleteByQueryBatchSize";
+    private final ClientProvider clientProvider;
+    private final ExecutorService executor;
+    private final int batchSize;
+
+    @Inject
+    public DeleteByQueryPerformer(ClientProvider clientProvider,
+                                  @Named(DELETE_BY_QUERY_POOL_SIZE) Integer threadPoolSize,
+                                  @Named(DELETE_BY_QUERY_BATCH_SIZE) Integer batchSize) {
+        checkValueIsPositive(threadPoolSize);
+        checkValueIsPositive(batchSize);
+        this.clientProvider = clientProvider;
+        this.executor = Executors.newFixedThreadPool(threadPoolSize);
+        this.batchSize = batchSize;
+    }
+
+    private void checkValueIsPositive(Integer value) {
+        Preconditions.checkNotNull(value);
+        Preconditions.checkArgument(value > 0, "value must be positive");
+    }
+
+    @PreDestroy
+    public void stop() {
+        executor.shutdownNow();
+    }
+
+    public void deleteByQuery(QueryBuilder queryBuilder) {
+        executor.execute(() -> doDeleteByQuery(queryBuilder));
+    }
+
+    private void doDeleteByQuery(QueryBuilder queryBuilder) {
+        try (Client client = clientProvider.get()) {
+            ListenableActionFuture<SearchResponse> searchResponseFuture = client.prepareSearch(ElasticSearchIndexer.MAILBOX_INDEX)
+                .setTypes(ElasticSearchIndexer.MESSAGE_TYPE)
+                .setScroll(new TimeValue(60000))
+                .setQuery(queryBuilder)
+                .setSize(batchSize)
+                .execute();
+
+            while(true) {
```

Comment:
I really don't like that : I would rather create a supplier that would supply some SearchResponse for each loop turn, or anything that would be higher level that what you did